packages/runtime/container-runtime/src/compatUtils.ts

```diff
@@ -0,0 +1,183 @@
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils/internal";
+import { FlushMode } from "@fluidframework/runtime-definitions/internal";
+import * as semver from "semver";
+
+import {
+	type ICompressionRuntimeOptions,
+	type IContainerRuntimeOptionsInternal,
+} from "./containerRuntime.js";
+import { pkgVersion } from "./packageVersion.js";
+
+/**
+ * Our policy is to support N/N-1 compatibility by default, where N is the most recent public major release of the runtime.
+ * Therefore, if the customer does not provide a compatibility mode, we will default to use N-1.
+ *
+ * However, this is not consistent with today's behavior. Some options (i.e. batching, compression) are enabled by default
+ * despite not being compatible with 1.x clients. Since the policy was introduced during 2.x's lifespan, N/N-1 compatibility
+ * by **default** will be in effect starting with 3.0. Importantly though, N/N-1 compatibility is still guaranteed with the
+ * proper configurations set.
+ *
+ * TODO: This should be updated to "2.0.0" when 3.0 is released.
+ */
+export const defaultCompatibilityMode = "pre-3.0-default";
+
+/**
+ * Available compression algorithms for op compression.
+ * @legacy
+ * @alpha
+ */
+export enum CompressionAlgorithms {
+	lz4 = "lz4",
+}
+
+/**
+ * @legacy
+ * @alpha
+ */
+export const disabledCompressionConfig: ICompressionRuntimeOptions = {
+	minimumBatchSizeInBytes: Number.POSITIVE_INFINITY,
+	compressionAlgorithm: CompressionAlgorithms.lz4,
+};
+
+export const enabledCompressionConfig = {
+	// Batches with content size exceeding this value will be compressed
+	minimumBatchSizeInBytes: 614400,
+	compressionAlgorithm: CompressionAlgorithms.lz4,
+};
+
+/**
+ * Subset of the IContainerRuntimeOptionsInternal properties which are version-dependent.
+ *
+ * @remarks
+ * When a new option is added to IContainerRuntimeOptionsInternal, we must consider if it's a version-dependent option.
+ * If it's considered version-dependent, then a corresponding entry must be added to `versionDependentOptionConfigs`. If not, then
+ * it must be omitted from this type.
+ *
+ * See [TODO: DOC LINK] for more details.
+ */
+export type IContainerRuntimeOptionsVersionDependent = Required<
+	Omit<
+		IContainerRuntimeOptionsInternal,
+		| "chunkSizeInBytes"
+		| "maxBatchSizeInBytes"
+		| "loadSequenceNumberVerification"
+		| "summaryOptions"
+		| "compatibilityMode"
+	>
+>;
+
+/**
+ * Map of version-dependent IContainerRuntimeOptionsInternal keys to their compatibility related information.
+ * The key is the option name, and the value is an object containing:
+ * - minVersionRequired: The minimum version of the container runtime that is required to use the version-dependent option
+ * - legacyConfig: The default config of the option when the runtime does not meets the min version requirement
+ * - modernConfig: The default config of the option when the runtime meets the min version requirement
```

Comment:
Move these down to doc comments on the return type?  I looked for them there.

Also, some minor suggestions:
```suggestion
 * - minVersionRequired: The minimum version of the container runtime that is required to use the version-dependent option
 * - legacyConfig: The default config of the option when any collaborator's runtime does not meet the min version requirement
 * - modernConfig: The default config of the option when all collaborators' runtimes meet the min version requirement
```

Reply:
Maybe explain in the comment(s) that the "version" in question is the "compatibilityMode" runtime option